docs/page-objects/index.rst

```diff
@@ -0,0 +1,121 @@
+.. _page-objects:
+
+============
+Page objects
+============
+
+A page object is a code wrapper for a webpage that implements the logic to
+parse the raw webpage data into structured data.
+
+To use web-poet, :ref:`define page object classes <page-object-classes>` for
+your target websites, and :ref:`get the output item <output-item>` using a
+:ref:`web-poet framework <frameworks>`.
+
+.. _page-object-classes:
+
+Defining a page object class
+============================
+
+A page object class is a Python class that:
+
+-   Subclasses :class:`~web_poet.pages.Injectable` or is registered or decorated
+    with :class:`Injectable.register <web_poet.pages.Injectable.register>`.
+
+-   Declares :ref:`typed input parameters <inputs>` in its ``__init__`` method.
+
+-   Implements a ``to_item`` method, which can be synchronous or asynchronous,
+    and returns the webpage content as an :ref:`item <items>`.
+
+For example:
+
+.. literalinclude:: page.py
+
+
+Minimizing boilerplate
+----------------------
+
+There are a few ways for you to minimize boilerplate when defining a page
+object class.
+
+For example, you can use attrs_ to remove the need for a custom ``__init__``
+method:
+
+.. _attrs: https://www.attrs.org/en/stable/index.html
+
+.. literalinclude:: attrs.py
+
+Also, it is often best to subclass :class:`~web_poet.pages.ItemPage`, which
+subclasses :class:`~web_poet.pages.Injectable` and provides an implementation of the
+``to_item`` method based on :ref:`declared fields <fields>`:
+
+.. literalinclude:: itempage.py
+
+If your page object class needs :class:`~web_poet.page_inputs.HttpResponse` as
+input, there is also :class:`~web_poet.pages.WebPage`, an
+:class:`~web_poet.pages.ItemPage` subclass that declares an
+:class:`~web_poet.page_inputs.HttpResponse` input and provides helper methods
+to use it:
+
+.. literalinclude:: webpage.py
+
+
+.. _output-item:
+
+Getting the output item
+=======================
+
+You should :ref:`include your page object classes into a page object
+registry <rules-intro>`, e.g. decorate them with :func:`~.handle_urls`:
+
+.. literalinclude:: register.py
+
+Then, provided your page object class code is imported (see
+:func:`~web_poet.rules.consume_modules`), your :ref:`framework <frameworks>`
+can build the output item after you provide the required input, such as the
+target URL and the desired :ref:`output item class <items>`, as :ref:`shown in
```

Comment:
In the register.py example the item class is dict, so this doesn't work. I don't think it's intended that you can ask for a "dict" as an item, and get result from some page object which returns a dict.

Reply:
This works:

```python
from web_poet import WebPage, field, handle_urls
from web_poet.example import get_item


@handle_urls("books.toscrape.com")
class BookPage(WebPage[dict]):

    @field
    async def title(self):
        return self.css("h1::text").get()


item = get_item(
    "http://books.toscrape.com/catalogue/the-exiled_247/index.html",
    dict,
)
print(item)
# {'title': 'The Exiled'}
```

But it would make sense to me to explain at this point that, even though this works, it is a good practice to use something other than dict.

Reply:
I think it's an anti-pattern to use Page[dict]. Probably it'd be better to remove it from all examples. For example, consider dependency injection. If you have some parameter which is of type dict (e.g. an optional argument which user defind for some reason), it can be surprising that it's populated by some random page object - `dict` doesn't tell anything about the kind of data you'd like to get.

Reply:
https://github.com/scrapinghub/web-poet/pull/152/commits/946c8d7b72b701b96ba02c1477ef278ec0bdc488